core/trie/trie.go

```diff
@@ -192,16 +192,143 @@ func (t *Trie) Get(key *felt.Felt) (*felt.Felt, error) {
 	return &leafValue, nil
 }
 
+// Iterate the trie from startValue in ascending order until the consumer returned false or an error occur. Return true
+// if end of trie is reached.
+// TODO: its much more efficient to iterate from the txn level. But even without that, if the leaf are ordered correctly,
+// block cache should have a pretty good hit rate.
+func (t *Trie) Iterate(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	startValueKey := t.feltToKey(startValue)
+	neverStopped, err := t.doIterate(&startValueKey, t.rootKey, consumer)
+
+	return neverStopped, err
+}
+
+func (t *Trie) doIterate(startValue, key *Key, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	if key == nil {
+		return false, nil
+	}
+
+	thenode, err := t.storage.Get(key)
+	if err != nil {
+		return false, err
+	}
+
+	if key.Len() == t.height {
+		if startValue.CmpAligned(key) > 0 {
+			return true, nil
+		}
+		keyAsFelt := key.Felt()
+		return consumer(&keyAsFelt, thenode.Value)
+	}
+
+	// If the startvalue is higher than the right node, no point in going to left at all
+	if startValue.CmpAligned(thenode.Right) < 0 {
+		next, err := t.doIterate(startValue, thenode.Left, consumer)
+		if err != nil {
+			return false, err
+		}
+
+		if !next {
+			return false, nil
+		}
+	}
+
+	return t.doIterate(startValue, thenode.Right, consumer)
+}
+
+type ProofNode struct {
+	Key  *Key
+	Hash *felt.Felt
+}
```

Comment:
We need to come up with a standard way of representing a proof node across different implementation.
I am not sure if pathfinder will be able to consume this proof.

We probably need to use the definition from the spec. https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/#specifications

What spec defines as `path` is a bit different. We calculate that in `path(key, parentKey *Key) Key`

Reply:
https://github.com/starknet-io/starknet-p2p-specs/pull/8/files#diff-1895f60c4bbe4eec02ba055cd529fd9ad7cb4b76bad983cbc68a2d1dc0ce0190R6-R21

Reply:
The spec specification for path seems to be for calculating the commitment, not the actual tree path. You can assume this as `JunoSnapProofNode` if you want and rename it later. As for the p2p spec, I thought we would rather make it juno specific first. Or would you rather have me dump the whole snap prototype branch and you guys can figureout how to implement it?

Reply:
I'll assume you'll adapt it accordingly.